lara-typescript/src/interactive-api-client/types.ts

```diff
@@ -297,6 +299,21 @@ export interface ICloseModal {
   uuid?: string;
 }
 
+export interface IDecoratedContentEvent {
+  type: string;
+  text: string;
+  bounds?: DOMRect;
+}
+
+export interface ITextDecorationInfo {
+  words: string[];
+  replace: string;
+  wordClass: string;
+  eventListeners: any;
```

Comment:
This looks better, but what about these listeners?

If I understand this right, an interactive would add a ITextDecorationHandler to the client to be notified when the host (AP) wants to decorate content. So then I'd guess the eventListeners here are a kind of 'proxy listener' which need to be notified when the user clicks (or whatever action) on the decorated content. It is a 'proxy listener' because what it does is forward this click action up to the host (AP). 

So it seems to good that the listeners here should be typed too. So some interactive which supports decorated content knows what it should send to these listeners.

Reply:
Good point.  When the AP passes this down, we really only care about the `type` that is inside of the object.  As discussed elsewhere, we can't actually pass functions through post message so we're actually adding the listeners locally based on the type that we get inside of the `eventListeners` object.  I'm thinking that what this really should be is something like:
`listenerTypes: string[]`
At present all we handle is "click", but we could imagine a `listenerTypes` array also containing a type for hover or active or double click or whatever else someone wants to do.  It would be up to the client to create the listener, but it would know that if it sent a message back up to the server when the event gets handled, that the server would do something with it.

Reply:
Although maybe we should preserver the function so the type can be exported and used elsewhere.  Not sure.  It would have to be optional.

Reply:
Although since this type is being created here for use in the messaging, adding a listener function (that will always be empty) seems wrong.  Again, I think it should just be an array of strings that represent what event types are supported:
`listenerTypes: string[]`
Thoughts?

Reply:
The goal is to make this as easy as possible for interactives to use. This is a client library which interactives include to reduce the work they need to do to support the interactive api. 

If the interactives are going to be using https://github.com/concord-consortium/text-decorator/blob/master/src/text-decorator.ts to do the actual decoration. Then matching these types up with what that library needs seems best. 
However this seems hard since that library supports different options: HTML or React. What the interactive needs to pass to the library varies depending on what kind of content the interactive is decorating.

In any case it is possible to have the eventListeners as functions here since interactive-api-client is passing this info. So the interactive-api-client could create a listener function that way the interactive wouldn't have to do that itself.  I don't know if that really makes since. It would depend if that makes it easier or harder for interactives to use this.

Basically I'm suggesting that this listener be moved into the interactive-api-client: https://github.com/concord-consortium/lara/blob/035d13ca19693d1ecf2450814a5deba2dd6cf3aa/lara-typescript/src/example-interactive/src/components/runtime.tsx#L42-L54

But I realize that is probably not going to be helpful for an interactive that is using React components instead of html. I haven't yet looked at what your changes look like in question interactives.

Reply:
I haven't pushed anything to question-interactives yet, but I was basically adding what we see above: a listener function that gets created in question-interactives.  I think I need clarification on what you want me to do next.  You want me to keep the `eventListeners` in the `ITextDecorationInfo` type?  And then move the creation of the handler and the decoration into the interactive-api-client rather than having client handle the message and decide what to do with it?

Reply:
"If the interactives are going to be using https://github.com/concord-consortium/text-decorator/blob/master/src/text-decorator.ts to do the actual decoration. Then matching these types up with what that library needs seems best."
Which seems a little odd since the `ITextDecorationInfo ` type is being used to type a message from the AP that can't have a function, but the types used in the TextDecorator calls require functions.
https://github.com/concord-consortium/text-decorator/blob/master/src/text-decorator.ts#L96
So you want this?
```
interface IEventListener {
  type: string;
  listener: (evt: Event) => void;
}

type IEventListeners = IEventListener | IEventListener[];

export interface ITextDecorationInfo {
  words: string[];
  replace: string;
  wordClass: string;
  eventListeners: IEventListeners;
}
```

Which then would not work with the AP unless `listener` is optional in `IEventListener` (or the AP passes down a stub function?).

Reply:
I am not understanding @scytacki - it seems like we have a working implementation that we could, in theory, refine further but at what cost? Is it really necessary to change the style of the implementation at this point?

Reply:
I'm not talking about changing the style. Just moving something from one side of the abstraction line to the other. I'll make a quick branch off of this to demonstrate.

I think it would also be fine to go ahead with this as is. And we can try to refine it later. I just like to fix things when we can.  Once they get into production it is a lot hard to update them when they are at the intersection of several components.

Reply:
Here is what I was thinking: https://github.com/concord-consortium/lara/pull/666

Reply:
Quick follow up on types.  I did go ahead and type this:
```
export interface ITextDecorationInfo {
  words: string[];
  replace: string;
  wordClass: string;
  eventListeners: IEventListeners;
}
```
Which still feels wrong since `IEventListeners` is an `IEventListener` or an array of `IEventListener` objects which include `listener: (evt: Event) => void;` which is not what anyone can actually send down via post message since the function gets stripped out.  I still think this should be `listenerTypes: string[]` which would, for example, be something like `["click"]`.  Keeping it as an `IEventListeners` type seems very misleading.

Reply:
Yeah the type going over the wire (postMessage) seems better if was listenerTypes. It might be best for it to be an array of objects incase we want to add more details to each listener in the future so the type would be `listenerTypes: Array<{type: string}>`. I didn't do that in my suggestion because I was trying to minimize the changes to the various parts. 

This does mean there are 2 types one for what goes over the wire(postMessage).  And another for what is provided to useDecorateContent. I think that is fine to have those two types. I think we have something like that is some other parts of the API.

Reply:
Yes, I think you're right that we would need two types for this.  I'll look at that today.

Reply:
@scytacki also, if you get a chance, take a look at the current state of this PR.  I think the last round of commits are in line with your thinking from last Friday.

Reply:
Yes your changes look good. I'm going to resolve this conversation. If we need more conversation about the types. Lets start a new one.